bayes_kit/ensemble.py

```diff
@@ -1,68 +1,91 @@
-    #     if init != None and init.shape != self._drawshape:
-    #         raise ValueError(f"init must be shape of draw {self._drawshape}; found {init.shape=}")
-    #     self._thetas = init or np.random.normal(size=self._drawshape)
-    #     self._firsthalf = range(halfwalkers)
-    #     self._secondhalf = range(halfwalkers, walkers)
+    def __init__(
+            self,
+            model: LogDensityModel,
+            a: Optional[float] = None,
+            walkers: Optional[int] = None,
+            init: Optional[NDArray[np.float64]] = None
+        ):
+        """
+        Initialize the sampler with a log density model, and optionally
+        proposal bounds, number of walkers and initial parameter values.
+            
+        Parameters:
+        model: class used to evaluate log densities
+        a: bounds on proposal (default 2)
+        walkers: an even number of walkers to use (default dimensionality of `model * 2`)
+        init: `walker` x `dimensio`n array of initial positions (defaults to standard normal)
+
+        Throws:
+        ValueError: if `a` is provided and not >= 1, `walker`s is provided and not strictly positive and even,
+        or if the `init` is provided and is not an `NDArray` of shape `walker` x `dimension`
+        """
+        self._model = model
+        self._dim = self._model.dims()
+        if a != None and np.float64(a) < 1:
+            raise ValueError(f"stretch bound must be greater than or equal to 1; found {a=}")
+        self._a = np.float64(a or 2.0)
+        self._sqrt_a = np.sqrt(np.float64(a))
+        self._inv_sqrt_a = 1 / self._sqrt_a
+        self._walkers = np.int64(walkers or 2 * self._dim)
+        if self._walkers < 2 or self._walkers % 2 != 0:
+            raise ValueError(f"walkers must be strictly positive, even integer; found {walkers=}")
+        self._halfwalkers = self._walkers // 2
+        self._drawshape = (int(self._walkers), self._dim)
+        self._thetas = np.asarray(init or np.random.normal(size=self._drawshape))
+        if self._thetas.shape != self._drawshape:
+            raise ValueError(f"init must be shape of draw {self._drawshape}; found {self._thetas.shape=}")
+        self._firsthalf = range(0, int(self._halfwalkers))
+        self._secondhalf = range(int(self._halfwalkers), int(self._walkers))
 
-    # def __iter__(self):
-    #     return self
+    def __iter__(self) -> Iterator[Sample]:
+        return self
 
-    # def __next__(self):
-    #     return self.sample
+    def __next__(self) -> Sample:
+        return self.sample()
 
-    # def draw_z(self):
-    #     """Return random draw z in (1/a, a) with p(z) propto 1 / sqrt(z)"""
-    #     return np.square(np.random.uniform(self._inv_sqrt_a, self._sqrt_a))
+    def draw_z(self) -> Sample:
+        """Return random draw z in (1/a, a) with p(z) propto 1 / sqrt(z)"""
+        return np.asarray(np.square(np.random.uniform(self._inv_sqrt_a, self._sqrt_a)))
```

Comment:
this np.asarray call is unneccessary. I assume it is there to make mypy happy, in which case you can do something like:

```python
draw: NDArray[np.float64] = np.square(np.random.uniform(self._inv_sqrt_a, self._sqrt_a))
return draw
```

Reply:
Thanks.